### PR TITLE
Firstcal update

### PR DESCRIPTION
**Pull request by AaronParsons**
This pull request re-implements firstcal within the redcal/linsolve framework.  If accepted, this PR should fully deprecate everything in firstcal.py, omni.py, and the omnical package.

this PR does the following:
- adds the RedundantCalibrator.firstcal method
- adds utils.fft_dly which is a simplified version of abscal_funcs.fft_dly.  For the future, we should try to sell Nick on migrating abscal code to using this function. **[Josh's Edit: I've made an issue for @nkern: #367]**
- refactors RedundantCalibrator.remove_degen into two functions.  One operates on phases/delays and does not require visibility solutions (only antenna solutions), and one operates on complex data and requires both antenna and visibility solutions.
- adds testing for firstcal, fft_dly, remove_degen, and remove_degen_gains

### CODE REVIEW

**Comment by AaronParsons** [comment]
I'm not totally happy with the RedundantCalibrator.firstcal interface.  It currently requires "df", the delta frequency between bins needed to set the delay scaling, and it returns antenna delays (not complex gains).  This interface is different from logcal/omnical/lincal, so it breaks parallels with that code a bit.  

The reason for the broken parallel is that degeneracy projection of firstcal solutions is best done on the delays themselves (as opposed to the phases at each frequency) to avoid introducing unwanted spectral structure.  In order to support this, it seems to make the most sense to have firstcal return the delays; re-computing phases from complex gains would be annoying and somewhat fraught.  Unfortunately, this means that firstcal needs to know the scale of the frequency axis to return delays in a sensible format.  Hence, df.

An option I did not implement was to embed degeneracy removal into firstcal.  This would allow the removal of the 'df' option and would allow firstcal to return complex gains instead of delays.  However, if you wanted to provide a degen_sol for this degeneracy solution, that degen_sol would have to be in delay, not gains, which would be ugly.  On the other hand, firstcal usually doesn't have a degen_sol to start with, so maybe we could just remove that option, in which case this would probably be the cleanest option.

If we go this route, we should also embed remove_degen into logcal/lincal/omnical to have a unified interface.  This change would need to be propagated into pipeline code as well.  I'm a bit ambivalent about it, so I haven't implemented these changes. 

As stands, the code is functional and usable, but maybe a bit annoying for the pipeline coder, since they need to manually convert delays into gains for firstcal.  They also need to call remove_degen manually for firstcal (using remove_degen_gains) and for any following logcal/lincal/omnical steps (using remove_degen).

**Comment by jsdillon** [comment]
I was going to comment that this PR is missing one of the big features of Zaki's `firstcal`, which was the post-FFT linear fit of relative delays. But then I figured I'd just test it and see if that feature is necessary.

At first blush, it seems like it is: I was getting pretty big deviations between new and old `firstcal`, not to mention large variations in delay from integration to integration:

screen shot 2018-09-20 at 5 20 34 pm

This gets even worse when I take off `medfilt`:
screen shot 2018-09-20 at 5 21 21 pm

This turned out to be a clue. Since `medfilt` is support to mitigate RFI, I was reminded of how Zaki solved this problem: he divided by the absolute value:

https://github.com/HERA-Team/hera_cal/blob/f4c742de572585c171a5a494ee253825c2ecb820/hera_cal/firstcal.py#L74

Since we're only fitting phase, this brings RFI down to the same level as signal, which makes the fit quite good:

screen shot 2018-09-20 at 5 18 23 pm

So good, in fact, that we might not need `medfilt` at all:

screen shot 2018-09-20 at 5 17 02 pm

Long story short, after `d12 = d1 * np.conj(data[bl2])` you should add `d12 /= np.where(np.abs(d12) == 0., 1., np.abs(d12))`.

I'd also consider removing the interface to `medfilt` in firstcal or, at the very least, making it default to `False`.

**Comment by AaronParsons** [comment]
> Long story short, after d12 = d1 * np.conj(data[bl2]) you should add d12 /= np.where(np.abs(d12) == 0., 1., np.abs(d12)).
> I'd also consider removing the interface to medfilt in firstcal or, at the very least, making it default to False.

@jsdillon, it might not surprise you that I was the original inventor of the "divide by abs" approach :).  It is robust, but for well-flagged data, also less sensitive.  I was curious whether medfilt was an adequate replacement.  Seems not.  Did you/can you put these tests into the unittests so we have exposure to medfilt's failings?

**Comment by jsdillon** [comment]
At least in the current pipeline, we don't expect firstcal to have any flags going in. Perhaps we want to have medfilt be default false and divide by abs be default true? And maybe a warning if neither is enabled?

I'm not entirely sure how to unit test this without introducing a large data file. Also, My metric has been consistency with old firstcal (and by proxy, from integration to integration). But we want to deprecate all that code...

File: hera_cal/redcal.py [hunk 1]
```diff
@@ -507,6 +507,52 @@ def compute_ubls(self, data, gain_sols):
...
+            sparse: represent the A matrix (visibilities to parameters) sparsely in linsolve
+            medfilt : boolean, median filter data before fft
+            kernel : size of median filter kernel along (time, freq) axes
+
+        Returns:
+            sol: dictionary of per-antenna delay solutions in the {(index,antpol): np.array} format.
+        """
+        fc_data = DataContainer(deepcopy(data))
+        calibrate_in_place(fc_data, sol0)
+        Nfreqs = fc_data.values()[0].shape[1]  # hardcode freq is axis 1 (time is axis 0)
```
[the remark this answers is not in the document]
Comment (AaronParsons) [reply hera_cal/redcal.py]: Replaced with:
Nfreqs = six.next(six.itervalues(data)).shape[1]  # hardcode freq is axis 1 (time is axis 0)

File: hera_cal/redcal.py [hunk 2]
```diff
@@ -507,6 +507,52 @@ def compute_ubls(self, data, gain_sols):
...
         return ubl_sols
 
+    def firstcal(self, data, wgts={}, df=1., sparse=False, mode='default', medfilt=True, kernel=(1, 11)):
+        """Solves for a per-antenna delay by fitting a line to the phase difference between
+        nominally redundant measurements.
+
+        Args:
+            data: visibility data in the dictionary format {(ant1,ant2,pol): np.array}
+            wgts: dictionary of linear weights in the same format as data. Defaults to equal wgts.
+            df: frequency change between data bins, determines scale of delays returned.
```
[the remark this answers is not in the document]
Comment (AaronParsons) [reply hera_cal/redcal.py]: Done.

File: hera_cal/redcal.py [hunk 3]
```diff
@@ -507,6 +507,52 @@ def compute_ubls(self, data, gain_sols):
...
+            sparse: represent the A matrix (visibilities to parameters) sparsely in linsolve
+            medfilt : boolean, median filter data before fft
+            kernel : size of median filter kernel along (time, freq) axes
+
+        Returns:
+            sol: dictionary of per-antenna delay solutions in the {(index,antpol): np.array} format.
+        """
+        fc_data = DataContainer(deepcopy(data))
+        calibrate_in_place(fc_data, sol0)
```
[the remark this answers is not in the document]
Comment (AaronParsons) [reply hera_cal/redcal.py]: Fixed this and another type in test_utils.

File: hera_cal/redcal.py [hunk 4]
```diff
@@ -507,6 +507,52 @@ def compute_ubls(self, data, gain_sols):
...
+        Nfreqs = fc_data.values()[0].shape[1]  # hardcode freq is axis 1 (time is axis 0)
+        if len(wgts) == 0:
+            wgts = {k: 1. for k in data}
```
[the remark this answers is not in the document]
Comment (AaronParsons) [reply hera_cal/redcal.py]: Done.

File: hera_cal/redcal.py [hunk 5]
```diff
@@ -507,6 +507,52 @@ def compute_ubls(self, data, gain_sols):
...
+        for (bl1, bl2), tau_ij in taus.items():
+            ai, aj = split_bl(bl1)
+            am, an = split_bl(bl2)
+            i, j, m, n = (self.pack_sol_key(k) for k in (ai, aj, am, an))
+            eq_key = '%s-%s-%s+%s' % (i, j, m, n)
+            d_ls[eq_key] = np.array(tau_ij)
+            w_ls[eq_key] = twgts[(bl1, bl2)]
+        ls = linsolve.LinearSolver(d_ls, wgts=w_ls, sparse=sparse)
+        sol = ls.solve(mode=mode)
+        freqs = np.arange(Nfreqs) * df
```
[the remark this answers is not in the document]
Comment (AaronParsons) [reply hera_cal/redcal.py]: Done.  (and as an aside, if you aren't using the phase offset, the starting point in freq doesn't matter)

File: hera_cal/utils.py [hunk 6]
```diff
@@ -100,6 +101,48 @@ def make_bl(*args):
...
+    if wgts is None:
+        wgts = np.ones_like(data, dtype=np.float)
+    # smooth via median filter
+    if medfilt:
+        data.real = signal.medfilt(data.real, kernel_size=kernel)
+        data.imag = signal.medfilt(data.imag, kernel_size=kernel)
+    # fft w/ window and find argmax
+    vfft = np.fft.fft(data * wgts, axis=1)
+    amp = np.abs(vfft)
+    argmaxes = np.argmax(amp, axis=1)
```
[the remark this answers is not in the document]
Comment (AaronParsons) [reply hera_cal/utils.py]: Zeroed out nans.

File: hera_cal/redcal.py [hunk 7]
```diff
@@ -592,99 +638,114 @@ def omnical(self, data, sol0, wgts={}, gain=.3, conv_crit=1e-10, maxiter=50, che
...
-            sol: dictionary that contains both visibility and gain solutions in the
-                {(ind1,ind2,pol): np.array} and {(index,antpol): np.array} formats respectively
-            degen_sol: Optional dictionary in the same format as sol. Gain amplitudes and phases
+            gains: dictionary that contains gain solutions in the {(index,antpol): np.array} format.
+            degen_gains: Optional dictionary in the same format as gains. Gain amplitudes and phases
                 in degen_sol replace the values of sol in the degenerate subspace of redcal. If
                 left as None, average gain amplitudes will be 1 and average phase terms will be 0.
-                Visibilties in degen_sol are ignored. Putting in firstcal solutions here can
-                help avoid phasewrapping issues.
+                Putting in firstcal solutions here can help avoid phasewrapping issues.
```
[the remark this answers is not in the document]
Comment (AaronParsons) [reply hera_cal/redcal.py]: Done.

File: hera_cal/redcal.py [hunk 8]
```diff
@@ -592,99 +638,114 @@ def omnical(self, data, sol0, wgts={}, gain=.3, conv_crit=1e-10, maxiter=50, che
...
+                degenMeanSqAmplitude = np.mean([np.abs(degen_gains[k1] * degen_gains[k2]) for k1 in gains.keys()
+                                                for k2 in gains.keys() 
+                                                if k1[1] == pol and k2[1] == pol and k1[0] != k2[0]], axis=0)
+                gainSols[gainPols == pol] *= (degenMeanSqAmplitude / meanSqAmplitude)**.5
 
         # Create new solutions dictionary
         newSol = {ant: gainSol for ant, gainSol in zip(ants, gainSols)}
-        newSol.update({bl_pair: visSol for bl_pair, visSol in zip(bl_pairs, visSols)})
-
         return newSol
```
[the remark this answers is not in the document]
Comment (AaronParsons) [reply hera_cal/redcal.py]: Done.

File: hera_cal/tests/test_redcal.py [hunk 9]
```diff
@@ -282,6 +283,29 @@ def solver(data, wgts, **kwargs):
...
+        fc_gains = {i: v.reshape(1, NFREQ) for i, v in fc_gains.items()}
+        gains = {k: v * fc_gains[k] for k, v in g.items()}
+        calibrate_in_place(d, gains, old_gains=g, gain_convention='multiply')
+        w = dict([(k, 1.) for k in d.keys()])
+        sol = info.firstcal(d, df=fqs[1] - fqs[0], medfilt=False)
+        sol_degen = info.remove_degen_gains(antpos, sol, degen_gains=delays, mode='phase')
+        for i in xrange(NANTS):
+            #self.assertEqual(sol[(i, 'Jxx')].shape, (1, NFREQ))
+            self.assertEqual(sol[(i, 'Jxx')].shape, (1, 1))
+            self.assertAlmostEqual(2 * sol_degen[(i, 'Jxx')], 2 * delays[(i, 'Jxx')], 0)
```
[the remark this answers is not in the document]
Comment (AaronParsons) [reply hera_cal/tests/test_redcal.py]: Nope.  Not with individual delay estimates only accurate at the ~5 ns level.

File: hera_cal/tests/test_utils.py [hunk 10]
```diff
@@ -65,6 +66,78 @@ def test_make_bl(self):
...
+        np.testing.assert_almost_equal(5 * dlys, 5 * true_dlys, -1)  # accuracy of 2 ns
+        mdl = np.exp(2j * np.pi * self.freqs.reshape((1, -1)) * dlys + 1j * offs)
+        np.testing.assert_almost_equal(np.angle(data * mdl.conj()), 0, -1)
+
+    def test_noisy(self):
+        true_dlys = np.random.uniform(-200, 200, size=60)
+        true_dlys.shape = (60, 1)
+        data = np.exp(2j * np.pi * self.freqs.reshape((1, -1)) * true_dlys) + 5 * noise((60, 1024))
+        df = np.median(np.diff(self.freqs))
+        dlys, offs = utils.foffs, 0.123, -2)offs, 0.123, -2ting.assert_almost_equal(np.angle(data * mdl.conj()), 0, -1) ft_dly(data, df)
```
[the remark this answers is not in the document]
Comment (AaronParsons) [reply hera_cal/tests/test_utils.py]: git blame says this was your fault.  I fixed it. ;)

File: hera_cal/tests/test_redcal.py [hunk 11]
```diff
@@ -445,6 +469,35 @@ def test_svd_convergence(self):
...
+    def test_remove_degen_firstcal_1D(self):
+        pol = 'xx'
+        xhat = np.array([1., 0, 0])
+        dtau_dx = 10.
+        antpos = build_linear_array(10)
+        reds = om.get_reds(antpos, pols=[pol], pol_mode='1pol')
+        rc = om.RedundantCalibrator(reds)
+        # put in a linear slope in delays, see that it is taken out
+        true_dlys = {(i, split_pol(pol)[0]): np.array([[np.dot(xhat, antpos[i]) * dtau_dx]]) for i in range(len(antpos))}
+        dlys = rc.remove_degen_gains(antpos, true_dlys, mode='phase')
```
[the remark this answers is not in the document]
Comment (AaronParsons) [reply hera_cal/tests/test_redcal.py]: Added.

File: hera_cal/redcal.py [hunk 12]
```diff
@@ -507,6 +507,52 @@ def compute_ubls(self, data, gain_sols):
...
+            wgts = {k: 1. for k in data}
+        wgts = DataContainer(wgts)
+        taus, twgts = {}, {}
+        for bls in self.reds:
+            for i, bl1 in enumerate(bls):
+                d1, w1 = data[bl1], wgts[bl1]
+                for bl2 in bls[i + 1:]:
+                    d12 = d1 * np.conj(data[bl2])
+                    w12 = w1 * wgts[bl2]
+                    taus[(bl1, bl2)] = fft_dly(d12, df, wgts=w12, medfilt=medfilt, kernel=kernel)
```
[the remark this answers is not in the document]
Comment (AaronParsons) [reply hera_cal/redcal.py]: Nah, this was intentional.  Both behave the same way in the equations, so we can solve both at once.  I decide at the end whether to use the offset or not, and currently I do not, but I'd like to reserve the possibility of doing so someday.

File: hera_cal/redcal.py [hunk 13]
```diff
@@ -507,6 +507,52 @@ def compute_ubls(self, data, gain_sols):
             ubl_sols[blgrp[0]] = np.average(d_gp, axis=0)  # XXX add option for median here?
         return ubl_sols
 
+    def firstcal(self, data, wgts={}, df=1., sparse=False, mode='default', medfilt=True, kernel=(1, 11)):
```
[the remark this answers is not in the document]
Comment (AaronParsons) [reply hera_cal/redcal.py]: Ok.

File: hera_cal/utils.py [hunk 14]
```diff
@@ -100,6 +101,48 @@ def make_bl(*args):
...
+        a, b, c = amp[i, mx - 1], amp[i, mx], amp[i, (mx + 1) % Nfreqs]
+        if np.abs(a - 2 * b + c) > 0 and np.abs(a - c) > 0:
+            shift = .5 * (a - c) / (a - 2 * b + c)
+        else:
+            shift = 0
+        # use peak shift to linearly interpolate to get appropriate delay
+        dlys[i] = (1.0 - np.abs(shift)) * fftfreqs[mx] + np.abs(shift) * (fftfreqs[mx] + np.sign(shift) * dtau)
+    # Now that we know the slope, estimate the remaining phase offset
+    freqs = np.arange(Nfreqs) * df
+    freqs.shape = (1, -1)
```
[the remark this answers is not in the document]
Comment (AaronParsons) [reply hera_cal/utils.py]: okay

File: hera_cal/redcal.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (AaronParsons) [reply hera_cal/redcal.py]: Changed variable name to taus_offs
